### PR TITLE
[ISSUE #8877] Refactor lock in ReceiptHandleGroup  to make the lock can be properly released when future can not be completed.

### LINKED ISSUE #8877: [Bug] Releasing lock in Future.whenComplete() may lead to the lock not being properly released when using ThreadPoolExecutor with DiscardOldestPolicy.
### Before Creating the Bug Report

- [X] I found a bug, not just asking a question, which should be created in [GitHub Discussions](https://github.com/apache/rocketmq/discussions).

- [X] I have searched the [GitHub Issues](https://github.com/apache/rocketmq/issues) and [GitHub Discussions](https://github.com/apache/rocketmq/discussions)  of this repository and believe that this is not a duplicate.

- [X] I have confirmed that this bug belongs to the current repository, not other repositories of RocketMQ.

### Runtime platform environment

any

### RocketMQ version

any

### JDK Version

any

### Describe the Bug

Releasing lock in Future.whenComplete() may lead to the lock not being properly released when using ThreadPoolExecutor with DiscardOldestPolicy.

look this function:

org.apache.rocketmq.proxy.common.ReceiptHandleGroup#computeIfPresent

### Steps to Reproduce

1. ConsumerProcessorExecutor's queue is full
2. Submit a new task to ConsumerProcessorExecutor and discard oldest inner renew task 
3. handleData.unlock() can not be reached
4. handle will never be removed

### What Did You Expect to See?

unlock  properly

### What Did You See Instead?

can not release lock 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by qianye1001**
### Which Issue(s) This PR Fixes

Fixes #8877

### CODE REVIEW

**Review by guyinyou — APPROVED** [review]
lgtm

**Comment by dingshuangxi888** [comment]
LGTM

File: proxy/src/main/java/org/apache/rocketmq/proxy/common/ReceiptHandleGroup.java [hunk 1]
```diff
@@ -98,22 +99,44 @@ public long getOffset() {
...
             } catch (InterruptedException e) {
-                return false;
+                return null;
             }
         }
 
-        public void unlock() {
+        public void unlock(long lockTimeMs) {
+            // if the lock is expired, we don't need to unlock it
+            if (System.currentTimeMillis() - lockTimeMs > ConfigurationManager.getProxyConfig().getLockTimeoutMsInHandleGroup() * 2) {
```
Comment (fuyou001) [inline proxy/src/main/java/org/apache/rocketmq/proxy/common/ReceiptHandleGroup.java:137]: add log

File: proxy/src/main/java/org/apache/rocketmq/proxy/common/ReceiptHandleGroup.java [hunk 2]
```diff
@@ -98,22 +99,44 @@ public long getOffset() {
...
+                boolean result = this.semaphore.tryAcquire(timeoutMs, TimeUnit.MILLISECONDS);
+                long currentTimeMs = System.currentTimeMillis();
+                if (result) {
+                    this.lastLockTimeMs.set(currentTimeMs);
+                    return currentTimeMs;
+                } else {
+                    // if the lock is expired, can be acquired again
+                    long expiredTimeMs = ConfigurationManager.getProxyConfig().getLockTimeoutMsInHandleGroup() * 3;
+                    if (currentTimeMs - this.lastLockTimeMs.get() > expiredTimeMs) {
+                        synchronized (this) {
```
Comment (fuyou001) [inline proxy/src/main/java/org/apache/rocketmq/proxy/common/ReceiptHandleGroup.java:121]: add log

**Review by fuyou001 — APPROVED** [review]

**Review by lizhimins — APPROVED** [review]

**Pull request merged**